README.rst
Keyholder
=========

keyholder provides a means of allowing a group of trusted users to use a
shared SSH identity without exposing the identity's private key.

The agent binds the socket at this address by default

::

    /run/keyholder/agent.sock (0666)

Before the shared SSH agent can be used, it must be armed by a user with
access to the private key. This can be done by running:

::

    $ /usr/sbin/keyholder arm

Users in the trusted group can use the shared agent by running:

::

    $ SSH_AUTH_SOCK=/run/keyholder/agent.sock ssh remote-host ...

License
-------

`Apache 2.0 <https://www.apache.org/licenses/LICENSE-2.0>`__


debian/changelog
keyholder (1.0-1) unstable; urgency=medium

  * Initial release. (Closes: T203003)

 -- Antoine Musso <hashar@free.fr>  Fri, 10 Apr 2020 23:39:27 +0200


keyholder/tests/test_lock.py
"""Lock class tests."""

from keyholder.crypto import SshLock


def test_ssh_lock():
    """Test various operations around an SshLock instance."""
    lock = SshLock()

    # pylint: disable=singleton-comparison

    # start with it being unlocked
    assert lock.is_locked() == False
    assert repr(lock) == "<SshLock: False>"

    # lock and test if locked
    assert lock.lock(b"passphrase") == True
    assert lock.is_locked() == True

    # wrong passphrase
    assert lock.unlock(b"wrong") == False
    assert lock.is_locked() == True

    # unlock and test if unlocked
    assert lock.unlock(b"passphrase") == True
    assert lock.is_locked() == False

    # unlock while unlocked
    assert lock.unlock(b"wrong") == False
    assert lock.is_locked() == False

    # lock while locked
    assert lock.lock(b"passphrase") == True
    assert lock.is_locked() == True
    assert lock.lock(b"again") == False
    assert lock.is_locked() == True


keyholder/tests/test_daemon.py
"""SSH daemon main functionality tests."""

import collections
import io
import logging

import nacl.signing
from Crypto.PublicKey import RSA
import pytest

from keyholder import daemon
from keyholder.daemon import AGENT_MAX_LEN, SshAgentProtocolError
from keyholder.protocol.types import SshBytes, SshRequestPublicKeySignature
from keyholder.protocol.agent import (
    SshAgentRequest,
    SshAgentRequestCode,
    SshAgentResponse,
    SshAgentResponseCode,
    SshAgentLock,
    SshAgentSignatureFlags,
)

# pylint: disable=redefined-outer-name


class SshAgentMockConfig:
    """Mock version of SshAgentConfig, just holding state."""

    # pylint: disable=too-few-public-methods
    def __init__(self):
        self.perms = {}


class SshAgentMockServer:
    """Mock version of SshAgentServer, just holding state."""

    # pylint: disable=too-few-public-methods
    def __init__(self):
        self.keys = collections.OrderedDict()
        self.config = SshAgentMockConfig()
        self.lock = daemon.SshLock()


class SshAgentMockHandler(daemon.SshAgentHandler):
    """Mock version of SshAgentHandler, replacing I/O operations."""

    # pylint: disable=super-init-not-called
    def __init__(self):
        self.rfile = io.BytesIO()
        self.wfile = io.BytesIO()
        self.setup()

    def setup(self):
        self.server = SshAgentMockServer()
        self.user, self.groups = "nobody", {"nogroup"}

    def su(self, user, groups=None):  # pylint: disable=invalid-name
        """su to a specified user (and optionally, group."""
        # pylint: disable=attribute-defined-outside-init
        self.user = user
        if groups:
            self.groups = groups

    def write_to_agent(self, msg):
        """mock-specific function: write input to the read buffer."""
        self.rfile.write(msg)
        self.rfile.seek(-len(msg), io.SEEK_END)
        # initialize a new write buffer for the response
        self.wfile = io.BytesIO()

    def read_from_agent(self):
        """mock-specific function: read from the write buffer."""
        # seek to 0 assuming we have a clean buffer
        self.wfile.seek(0)
        read = self.wfile.read()
        response = SshAgentResponse.parse(read)
        return response.code, response.message

    def communicate(self, msg):
        """mock-specific function: write then read."""
        self.write_to_agent(msg)
        self.handle()
        return self.read_from_agent()


@pytest.fixture(scope="function")
def handler():
    """A fixture for holding an instance of SshAgentMockHandler."""
    return SshAgentMockHandler()


@pytest.mark.parametrize(
    "msg, exception",
    [
        (b"", EOFError),
        (b"\x00\x00\x00\x00", SshAgentProtocolError),
        (b"\x00\x00\x00\x04abc", EOFError),
        (b"\x00\x00\x00\x04\x00\x00\x00\x00", SshAgentProtocolError),
        (SshBytes.build(SshBytes.build(b"abc")), SshAgentProtocolError),
        (
            SshBytes.build(SshBytes.build(b"a" * AGENT_MAX_LEN)),
            SshAgentProtocolError,
        ),
    ],
)
def test_recv_request_invalid(handler, msg, exception):
    """Test invalid requests."""
    with pytest.raises(exception):
        handler.write_to_agent(msg)
        handler.recv_request()


def test_recv_request(handler):
    """Test a valid request."""
    handler.write_to_agent(
        SshAgentRequest.build(
            {
                "code": SshAgentRequestCode.LOCK,
                "message": SshAgentLock.build(b"passphrase"),
            }
        )
    )
    code, message = handler.recv_request()
    assert code == SshAgentRequestCode.LOCK
    assert message == SshAgentLock.build(b"passphrase")


@pytest.mark.parametrize(
    "msg",
    [
        b"\x00\x00\x00\x00",
        b"\x00\x00\x00\x04\x00\x00\x00\x00",
        SshBytes.build(SshBytes.build(b"abc")),
        SshBytes.build(SshBytes.build(b"a" * (AGENT_MAX_LEN + 1))),
    ],
)
def test_handle_invalid(handler, caplog, msg):
    """Test invalid requests."""
    caplog.set_level(logging.INFO)
    handler.write_to_agent(msg)
    handler.handle()
    assert "Invalid request received" in caplog.text


def test_send_invalid(handler):
    """Attempt to send invalid messages."""
    with pytest.raises(SshAgentProtocolError):
        handler.send_response(666)


def test_send_ioerror(handler, caplog):
    """Attempt to send to a closed socket."""
    caplog.set_level(logging.INFO)
    handler.wfile.close()
    handler.send_response(SshAgentResponseCode.SUCCESS)
    assert "Response write failed" in caplog.text


def test_permissions(handler):
    """Test the various permissions functions."""
    handler.su("root")
    assert handler.is_superuser()
    assert handler.is_allowed("dummy")

    handler.su("nobody")
    assert not handler.is_superuser()


def test_handle_unimplemented(handler, caplog):
    """Test an unimplemented function."""
    caplog.set_level(logging.DEBUG)
    req_code = SshAgentRequestCode.V1_ADD_RSA_ID_CONSTRAINED
    v1_request = SshAgentRequest.build({"code": req_code, "message": b""})
    code, _ = handler.communicate(v1_request)
    assert code == SshAgentResponseCode.FAILURE
    assert "Request type %s not implemented" % req_code.name in caplog.text


def test_handle_lock_unlock(handler):
    """Test agent locking and unlocking."""
    lock_request = SshAgentRequest.build(
        {
            "code": SshAgentRequestCode.LOCK,
            "message": SshAgentLock.build(b"passphrase"),
        }
    )
    unlock_request = SshAgentRequest.build(
        {
            "code": SshAgentRequestCode.UNLOCK,
            "message": SshAgentLock.build(b"passphrase"),
        }
    )
    unlock_invalid_request = SshAgentRequest.build(
        {
            "code": SshAgentRequestCode.UNLOCK,
            "message": SshAgentLock.build(b"invalid"),
        }
    )

    # start with an unlocked state
    assert not handler.server.lock.is_locked()

    # try once with no privileges
    handler.su("nobody")
    code, _ = handler.communicate(lock_request)
    assert code == SshAgentResponseCode.FAILURE

    # then try again with the right privileges
    handler.su("root")
    code, _ = handler.communicate(lock_request)
    assert code == SshAgentResponseCode.SUCCESS
    assert handler.server.lock.is_locked()

    # then attempt to lock twice
    handler.su("root")
    code, _ = handler.communicate(lock_request)
    assert code == SshAgentResponseCode.FAILURE

    # ensure that no operations are allowed, even for root
    handler.su("nobody")
    assert not handler.is_allowed("dummy")
    handler.su("root")
    assert not handler.is_allowed("dummy")

    handler.su("nobody")
    # try unlocking as a non-superuser
    code, _ = handler.communicate(unlock_request)
    assert code == SshAgentResponseCode.FAILURE
    assert handler.server.lock.is_locked()

    handler.su("root")
    # try first with an invalid passphrase
    code, _ = handler.communicate(unlock_invalid_request)
    assert code == SshAgentResponseCode.FAILURE
    assert handler.server.lock.is_locked()

    # then try again with the right passphrase + user, i.e. actually unlock
    code, _ = handler.communicate(unlock_request)
    assert code == SshAgentResponseCode.SUCCESS
    assert not handler.server.lock.is_locked()

    # try unlocking the unlocked agent again
    code, _ = handler.communicate(unlock_request)
    assert code == SshAgentResponseCode.FAILURE
    assert not handler.server.lock.is_locked()


def test_add_identity_invalid(handler):
    """Test adding invalid identities."""
    # first try to add an invalid key type
    ssh_add_identity_invalid = SshAgentRequest.build(
        {
            "code": SshAgentRequestCode.ADD_IDENTITY,
            "message": {"key_type": "ssh-invalid", "key": b"invalid"},
        }
    )

    handler.su("root")
    code, _ = handler.communicate(ssh_add_identity_invalid)
    assert code == SshAgentResponseCode.FAILURE

    # then an invalid ed25519 key
    ssh_add_identity_invalid = SshAgentRequest.build(
        {
            "code": SshAgentRequestCode.ADD_IDENTITY,
            "message": {
                "key_type": "ssh-ed25519",
                "key": {
                    "enc_a": b"invalid",
                    "k_enc_a": b"invalid",
                    "comment": "comment-invalid",
                },
            },
        }
    )
    handler.su("root")
    code, _ = handler.communicate(ssh_add_identity_invalid)
    assert code == SshAgentResponseCode.FAILURE


def test_add_identity_rsa(handler):
    """Test adding an RSA identity."""
    # generate an RSA key
    rsa_key = RSA.generate(2048)

    # then add the key to the agent
    ssh_add_identity = SshAgentRequest.build(
        {
            "code": SshAgentRequestCode.ADD_IDENTITY,
            "message": {
                "key_type": "ssh-rsa",
                "key": {
                    "n": rsa_key.n,
                    "e": rsa_key.e,
                    "d": rsa_key.d,
                    "iqmp": rsa_key.u,
                    "p": rsa_key.p,
                    "q": rsa_key.q,
                    "comment": "comment",
                },
            },
        }
    )
    # first try as a non-privileged user (and fail)
    handler.su("nobody")
    code, _ = handler.communicate(ssh_add_identity)
    assert code == SshAgentResponseCode.FAILURE

    # then actually add them to the agent
    handler.su("root")
    code, _ = handler.communicate(ssh_add_identity)
    assert code == SshAgentResponseCode.SUCCESS


def test_add_identity_ed25519(handler):
    """Test adding an Ed25519 identity."""
    # generate an Ed25519 key
    private_key = nacl.signing.SigningKey.generate()
    k = private_key.encode()
    enc_a = private_key.verify_key.encode()
    k_enc_a = k + enc_a

    # then add the key to the agent
    ssh_add_identity = SshAgentRequest.build(
        {
            "code": SshAgentRequestCode.ADD_IDENTITY,
            "message": {
                "key_type": "ssh-ed25519",
                "key": {
                    "enc_a": enc_a,
                    "k_enc_a": k_enc_a,
                    "comment": "comment",
                },
            },
        }
    )
    # first try as a non-privileged user (and fail)
    handler.su("nobody")
    code, _ = handler.communicate(ssh_add_identity)
    assert code == SshAgentResponseCode.FAILURE

    # then actually add them to the agent
    handler.su("root")
    code, _ = handler.communicate(ssh_add_identity)
    assert code == SshAgentResponseCode.SUCCESS


def test_add_remove_identities(handler):
    """Test adding, listing and removing identities."""
    ssh_request_identities = SshAgentRequest.build(
        {"code": SshAgentRequestCode.REQUEST_IDENTITIES, "message": b""}
    )

    handler.su("root")
    code, message = handler.communicate(ssh_request_identities)
    assert code == SshAgentResponseCode.IDENTITIES_ANSWER
    assert len(message) == 0  # pylint: disable=len-as-condition

    # now add a few keys to the agent
    n_keys = 3
    for key_id in range(n_keys):
        # generate an Ed25519 key
        private_key = nacl.signing.SigningKey.generate()
        k = private_key.encode()
        enc_a = private_key.verify_key.encode()
        k_enc_a = k + enc_a

        # then add this key to the agent
        ssh_add_identity = SshAgentRequest.build(
            {
                "code": SshAgentRequestCode.ADD_IDENTITY,
                "message": {
                    "key_type": "ssh-ed25519",
                    "key": {
                        "enc_a": enc_a,
                        "k_enc_a": k_enc_a,
                        "comment": "comment%d" % key_id,
                    },
                },
            }
        )
        # first try as a non-privileged user (and fail)
        handler.su("nobody")
        code, _ = handler.communicate(ssh_add_identity)
        assert code == SshAgentResponseCode.FAILURE

        # then actually add them to the agent
        handler.su("root")
        code, _ = handler.communicate(ssh_add_identity)
        assert code == SshAgentResponseCode.SUCCESS

    # re-add the last added key -- this, combined with the n_keys check below,
    # makes sure that duplicates are suppressed
    handler.su("root")
    code, _ = handler.communicate(ssh_add_identity)
    assert code == SshAgentResponseCode.SUCCESS

    # list identities, but with no permissions (i.e. should get empty list)
    handler.su("nobody")
    code, _ = handler.communicate(ssh_request_identities)
    assert code == SshAgentResponseCode.IDENTITIES_ANSWER
    assert len(message) == 0  # pylint: disable=len-as-condition

    # list identities, but with superuser permissions
    handler.su("root")
    code, message = handler.communicate(ssh_request_identities)
    assert code == SshAgentResponseCode.IDENTITIES_ANSWER
    assert len(message) == n_keys
    assert message[0].comment == "comment0"

    # now, try to remove the first identity from the agent
    key_blob_to_delete = message[0].key_blob
    ssh_remove_identity = SshAgentRequest.build(
        {
            "code": SshAgentRequestCode.REMOVE_IDENTITY,
            "message": {"key_blob": key_blob_to_delete},
        }
    )
    handler.su("nobody")
    code, _ = handler.communicate(ssh_remove_identity)
    assert code == SshAgentResponseCode.FAILURE

    handler.su("root")
    code, _ = handler.communicate(ssh_remove_identity)
    assert code == SshAgentResponseCode.SUCCESS

    # ...and verify that it was deleted
    handler.su("root")
    code, message = handler.communicate(ssh_request_identities)
    assert code == SshAgentResponseCode.IDENTITIES_ANSWER
    assert len(message) == n_keys - 1
    assert message[0].comment == "comment1"

    # try removing the same key again (should fail)
    handler.su("root")
    code, _ = handler.communicate(ssh_remove_identity)
    assert code == SshAgentResponseCode.FAILURE


def test_remove_all_identities(handler):
    """Verify that removing all identities works."""
    # add a key or two
    test_add_identity_rsa(handler)
    test_add_identity_ed25519(handler)

    # then try to remove all identities from the agent
    ssh_remove_all_identities = SshAgentRequest.build(
        {"code": SshAgentRequestCode.REMOVE_ALL_IDENTITIES, "message": b""}
    )
    handler.su("nobody")
    code, _ = handler.communicate(ssh_remove_all_identities)
    assert code == SshAgentResponseCode.FAILURE

    handler.su("root")
    code, _ = handler.communicate(ssh_remove_all_identities)
    assert code == SshAgentResponseCode.SUCCESS

    # and verify that they're all gone
    handler.su("root")
    ssh_request_identities = SshAgentRequest.build(
        {"code": SshAgentRequestCode.REQUEST_IDENTITIES, "message": b""}
    )
    code, message = handler.communicate(ssh_request_identities)
    assert code == SshAgentResponseCode.IDENTITIES_ANSWER
    assert len(message) == 0  # pylint: disable=len-as-condition


def test_sign_request(handler):
    """Test that signature requests work."""
    test_add_identity_ed25519(handler)
    test_add_identity_rsa(handler)

    handler.su("root")
    ssh_request_identities = SshAgentRequest.build(
        {"code": SshAgentRequestCode.REQUEST_IDENTITIES, "message": b""}
    )
    code, message = handler.communicate(ssh_request_identities)
    assert code == SshAgentResponseCode.IDENTITIES_ANSWER
    assert len(message) == 2

    for identity in message:
        key_blob = identity.key_blob
        ssh_signature = SshRequestPublicKeySignature.build(
            {
                "session": b"dummy-session",
                "username": "dummy-username",
                "servicename": "dummy-servicename",
                "algo": "ssh-ed25519",
                "key_blob": key_blob,
            }
        )
        ssh_sign_request = SshAgentRequest.build(
            {
                "code": SshAgentRequestCode.SIGN_REQUEST,
                "message": {
                    "key_blob": key_blob,
                    "data": ssh_signature,
                    "flags": SshAgentSignatureFlags.AGENT_NO_FLAGS,
                },
            }
        )
        # ask for a sign request (but with a key that we don't have perms for)
        handler.su("nobody")
        code, _ = handler.communicate(ssh_sign_request)
        assert code == SshAgentResponseCode.FAILURE

        # now do the same, but with superuser perms
        handler.su("root")
        code, _ = handler.communicate(ssh_sign_request)
        assert code == SshAgentResponseCode.SIGN_RESPONSE

        # ask with an invalid signature request (OpenSSH accepts this!)
        ssh_invalid_sign_request = SshAgentRequest.build(
            {
                "code": SshAgentRequestCode.SIGN_REQUEST,
                "message": {
                    "key_blob": key_blob,
                    "data": b"invalid",
                    "flags": SshAgentSignatureFlags.AGENT_NO_FLAGS,
                },
            }
        )
        handler.su("root")
        code, _ = handler.communicate(ssh_invalid_sign_request)
        assert code == SshAgentResponseCode.FAILURE

    # try with a non-existing key
    ssh_signature = SshRequestPublicKeySignature.build(
        {
            "session": b"dummy-session",
            "username": "dummy-username",
            "servicename": "dummy-servicename",
            "algo": "ssh-ed25519",
            "key_blob": b"random-garbage",
        }
    )
    ssh_sign_request = SshAgentRequest.build(
        {
            "code": SshAgentRequestCode.SIGN_REQUEST,
            "message": {
                "key_blob": b"random-garbage",
                "data": ssh_signature,
                "flags": SshAgentSignatureFlags.AGENT_NO_FLAGS,
            },
        }
    )
    handler.su("root")
    code, _ = handler.communicate(ssh_sign_request)
    assert code == SshAgentResponseCode.FAILURE


@pytest.mark.skip(reason="Takes too long")
def test_agent_max_len(handler):
    """Add a lot of keys then lists them, to trigger AGENT_MAX_LEN."""
    n_keys = 4000
    for key_id in range(n_keys):
        k = key_id.to_bytes(32, "big")
        private_key = nacl.signing.SigningKey(k)
        enc_a = private_key.verify_key.encode()
        k_enc_a = k + enc_a

        # then add this key to the agent
        ssh_add_identity = SshAgentRequest.build(
            {
                "code": SshAgentRequestCode.ADD_IDENTITY,
                "message": {
                    "key_type": "ssh-ed25519",
                    "key": {
                        "enc_a": enc_a,
                        "k_enc_a": k_enc_a,
                        "comment": "comment%d" % key_id,
                    },
                },
            }
        )
        # then actually do it
        handler.su("root")
        code, _ = handler.communicate(ssh_add_identity)
        assert code == SshAgentResponseCode.SUCCESS

    ssh_request_identities = SshAgentRequest.build(
        {"code": SshAgentRequestCode.REQUEST_IDENTITIES, "message": b""}
    )

    code, _ = handler.communicate(ssh_request_identities)
    assert code == SshAgentResponseCode.FAILURE


keyholder/tests/test_argparse.py
"""Argument parse tests."""

import pytest
import keyholder.daemon


def test_parse_args_help(capsys):
    """Test whether --help returns usage and exits"""
    with pytest.raises(SystemExit) as exc:
        keyholder.daemon.parse_args(["--help"])

    out, _ = capsys.readouterr()
    assert exc.value.code == 0
    assert "usage: " in out


keyholder/crypto.py
"""
SSH public/private keys crypto implementation.

Copyright 2018 Wikimedia Foundation, Inc.
Copyright 2018 Faidon Liambotis <faidon@wikimedia.org>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY CODE, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import base64
import binascii
import hashlib
import os

# for RSA
from Crypto.Hash import SHA, SHA256, SHA512
from Crypto.PublicKey import RSA
from Crypto.Signature import PKCS1_v1_5

# for ed25519
import nacl.signing

from keyholder.protocol.agent import SshAgentSignatureFlags
from keyholder.protocol.types import SshKeyBlob

# pylint: disable=no-else-return


def ssh_fingerprint(blob, hash_type='sha256'):
    """Returns the fingerprint of a key blob, in OpenSSH format."""
    if hash_type == 'md5':
        digest = hashlib.md5(blob).digest()
        hexdigest = b':'.join([binascii.hexlify(bytes([b])) for b in digest])
        return 'MD5:' + hexdigest.decode('ascii')
    elif hash_type == 'sha256':
        digest = hashlib.sha256(blob).digest()
        b64 = base64.b64encode(digest)
        return 'SHA256:' + b64.rstrip(b'=').decode('ascii')
    else:
        raise TypeError('Unrecognized fingerprint type %s' % hash_type)


class SshLock:
    """An instance of a lock, with lock() and unlock() operations."""
    def __init__(self):
        self.salt = os.urandom(16)
        self.iter = 100 * 1000
        self.locked = None

    def _pbkdf2(self, passphrase):
        """Helper function to return the PBKDF2 of a passphrase."""
        return hashlib.pbkdf2_hmac('sha256', passphrase, self.salt, self.iter)

    def is_locked(self):
        """Returns True if object is locked, False otherwise."""
        return self.locked is not None

    def lock(self, passphrase):
        """Lock the object with passphrase.

        Returns True if object was successfully locked.
        Returns False if object was already locked.
        """
        if not self.is_locked():
            self.locked = self._pbkdf2(passphrase)
            return True
        else:
            return False

    def unlock(self, passphrase):
        """Unlock the object with passphrase.

        Returns False if object was not already locked.
        Returns False if passphrase did not match.
        Returns True if object was successfully unlocked.
        """
        if not self.is_locked():
            return False

        attempted = self._pbkdf2(passphrase)
        if attempted == self.locked:
            self.locked = None
            return True
        else:
            return False

    def __repr__(self):
        return '<%s: %s>' % (self.__class__.__name__, self.is_locked())


class SshBaseKey:
    """Base class to represents an SSH key."""
    @property
    def fingerprint(self, hash_type='sha256'):
        """Returns the fingerprint of the current public key."""
        return ssh_fingerprint(self.key_blob, hash_type)

    def __repr__(self):
        return '<%s: %s>' % (self.__class__.__name__, self.fingerprint)

    @property
    def key_blob(self):
        """Returns the key blob of the public key, in OpenSSH format."""
        raise NotImplementedError

    def sign(self, data, flags):
        """Signs `data` with the private key."""
        raise NotImplementedError


class SshRSAKey(SshBaseKey):
    """SSH RSA key."""
    def __init__(self, tup, comment):
        super().__init__()
        self.private_key = RSA.construct(tup)
        self.comment = comment

    @property
    def key_blob(self):
        return SshKeyBlob.build({
            'algo': 'ssh-rsa',
            'blob': {
                'e': self.private_key.e,
                'n': self.private_key.n,
            },
        })

    def sign(self, data, flags):
        if flags == SshAgentSignatureFlags.AGENT_RSA_SHA2_256:
            hsh = SHA256.new(data)
        elif flags == SshAgentSignatureFlags.AGENT_RSA_SHA2_512:
            hsh = SHA512.new(data)
        else:
            hsh = SHA.new(data)

        signer = PKCS1_v1_5.new(self.private_key)
        signature = signer.sign(hsh)

        return {
            'key_type': 'ssh-rsa',
            'signature': signature,
        }


class SshEd25519Key(SshBaseKey):
    """SSH Ed25519 key."""
    def __init__(self, enc_a, k_enc_a, comment):
        super().__init__()
        k = k_enc_a[:32]
        if k + enc_a != k_enc_a:
            raise TypeError('Invalid key: ENC(A) does not match k || ENC(A)')

        self.private_key = nacl.signing.SigningKey(k)
        self.public_key = self.private_key.verify_key
        self.comment = comment

        if self.public_key.encode() != enc_a:
            raise TypeError('Invalid key: ENC(A) does not match k')

    @property
    def key_blob(self):
        return SshKeyBlob.build({
            'algo': 'ssh-ed25519',
            'blob': {
                'public_key': self.public_key.encode(),
            },
        })

    def sign(self, data, flags):
        signature = self.private_key.sign(data).signature
        return {
            'key_type': 'ssh-ed25519',
            'signature': signature,
        }


keyholder/protocol/types.py
"""
Basic SSH protocol data types

As defined in RFCs 4251, 4252, 4253 etc.

Copyright 2018 Wikimedia Foundation, Inc.
Copyright 2018 Faidon Liambotis <faidon@wikimedia.org>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY CODE, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

# pylint: disable=invalid-name

from construct import (
    Byte,
    Flag,
    Int32ub,
    PascalString,
    BytesInteger,
    GreedyBytes,
    Struct,
    FocusedSeq,
    Prefixed,
    Select,
    Switch,
    Rebuild,
    Terminated,
    this
)

from .compat import Const

# RFC 4251 section 5
SshBytes = Prefixed(Int32ub, GreedyBytes)
SshString = PascalString(Int32ub, 'utf8')
SshMPInt = Select(
    Const(0, Int32ub),  # zero stored as zero bytes of data
    FocusedSeq(
        'num',
        'len' / Rebuild(Int32ub,
                        lambda ctx: int(ctx.num.bit_length() // 8 + 1)),
        'num' / BytesInteger(this.len, signed=True),
    ),
)

# RFC 4253 section 6.6
SshDSAKeyBlob = Struct(
    'p' / SshMPInt,
    'q' / SshMPInt,
    'g' / SshMPInt,
    'y' / SshMPInt,
    Terminated
)

SshRSAKeyBlob = Struct(
    'e' / SshMPInt,
    'n' / SshMPInt,
    Terminated
)

# RFC 5656, section 3.1
SshECDSAKeyBlob = Struct(
    'identifier' / SshString,
    'Q' / SshBytes,
    Terminated
)

# I-D.ietf-curdle-ssh-ed25519, section 4
SshEd25519KeyBlob = Struct(
    'public_key' / SshBytes,
    Terminated
)

# RFC 4253 section 6.6
SshKeyBlob = Struct(
    'algo' / SshString,
    'blob' / Switch(
        this.algo, {
            'ssh-dss': SshDSAKeyBlob,
            'ssh-rsa': SshRSAKeyBlob,
            'ecdsa-sha2-nistp256': SshECDSAKeyBlob,
            'ecdsa-sha2-nistp384': SshECDSAKeyBlob,
            'ecdsa-sha2-nistp521': SshECDSAKeyBlob,
            'ssh-ed25519': SshEd25519KeyBlob,
        },
        default=GreedyBytes
    ),
    Terminated
)

# RFC 4253 section 6.6
SshSignature = FocusedSeq(
    'signature',
    'signature' / Prefixed(Int32ub, Struct(
        'key_type' / SshString,
        'signature' / SshBytes,
        Terminated
    )),
    Terminated
)

# RFC 4252, section 7
SSH_MSG_USERAUTH_REQUEST = 50
SshRequestPublicKeySignature = Struct(
    'session' / SshBytes,
    'type' / Const(SSH_MSG_USERAUTH_REQUEST, Byte),
    'username' / SshString,
    'servicename' / SshString,
    'method' / Const('publickey', SshString),
    'has_signature' / Const(True, Flag),
    'algo' / SshString,
    'key_blob' / SshBytes,
    Terminated,
)


keyholder/protocol/agent.py
"""
SSH Agent Protocol

This is based on the draft RFC about the SSH Agent protocol
I-D.ietf-miller-ssh-agent (as of September 2018).

Earlier SSHv1 commands are listed for reference, but not intended to be used
anymore. They can be found here: <http://api.libssh.org/rfc/PROTOCOL.agent>

Copyright 2018 Wikimedia Foundation, Inc.
Copyright 2018 Faidon Liambotis <faidon@wikimedia.org>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY CODE, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

# pylint: disable=invalid-name

import enum

from construct import (
    Byte,
    Int32ub,
    GreedyBytes,
    Struct,
    FocusedSeq,
    Prefixed,
    Switch,
    Rebuild,
    Terminated,
    this,
    len_
)

from keyholder.protocol.compat import PyEnum
from keyholder.protocol.types import (
    SshBytes,
    SshString,
    SshMPInt,
    SshSignature
)


@enum.unique
class SshAgentRequestCode(enum.Enum):
    """SSH Agent request codes."""
    V1_REQUEST_RSA_IDENTITIES = 1
    V1_RSA_CHALLENGE = 3
    V1_ADD_RSA_IDENTITY = 7
    V1_REMOVE_RSA_IDENTITY = 8
    V1_REMOVE_ALL_RSA_IDENTITIES = 9
    REQUEST_IDENTITIES = 11
    SIGN_REQUEST = 13
    ADD_IDENTITY = 17
    REMOVE_IDENTITY = 18
    REMOVE_ALL_IDENTITIES = 19
    ADD_SMARTCARD_KEY = 20
    REMOVE_SMARTCARD_KEY = 21
    LOCK = 22
    UNLOCK = 23
    V1_ADD_RSA_ID_CONSTRAINED = 24
    ADD_ID_CONSTRAINED = 25
    ADD_SMARTCARD_KEY_CONSTRAINED = 26
    EXTENSION = 27


@enum.unique
class SshAgentResponseCode(enum.Enum):
    """SSH Agent response codes."""
    V1_RSA_IDENTITIES_ANSWER = 2
    V1_RSA_RESPONSE = 4
    FAILURE = 5
    SUCCESS = 6
    IDENTITIES_ANSWER = 12
    SIGN_RESPONSE = 14
    EXTENSION_FAILURE = 28


@enum.unique
class SshAgentSignatureFlags(enum.Enum):
    """AGENTC_SIGN_REQUEST flags.

    These are flags that can, in theory, be combined, but in practice they all
    conflict with each other, hence they are listed as an Enum here.
    """
    AGENT_NO_FLAGS = 0
    V1_AGENT_OLD_SIGNATURE = 1
    AGENT_RSA_SHA2_256 = 2
    AGENT_RSA_SHA2_512 = 4


SshAgentIdentities = FocusedSeq(
    'keys',
    'nkeys' / Rebuild(Int32ub, len_(this.keys)),
    'keys' / Struct(
        'key_blob' / SshBytes,
        'comment' / SshString,
    )[this.nkeys],
    Terminated,
)

SshAddRSAKey = Struct(
    'n' / SshMPInt,
    'e' / SshMPInt,
    'd' / SshMPInt,
    'iqmp' / SshMPInt,
    'p' / SshMPInt,
    'q' / SshMPInt,
    'comment' / SshString,
)

SshAddEd25519Key = Struct(
    'enc_a' / SshBytes,
    'k_enc_a' / SshBytes,
    'comment' / SshString,
)

SshAddIdentity = Struct(
    'key_type' / SshString,
    'key' / Switch(
        this.key_type, {
            'ssh-ed25519': SshAddEd25519Key,
            'ssh-rsa': SshAddRSAKey,
        },
        default=GreedyBytes,
    ),
    Terminated,
)

SshRemoveIdentity = Struct(
    'key_blob' / SshBytes,
    Terminated
)

SshAgentSignRequest = Struct(
    'key_blob' / SshBytes,
    'data' / SshBytes,
    'flags' / PyEnum(Int32ub, SshAgentSignatureFlags),
    Terminated
)

SshAgentLock = SshAgentUnlock = FocusedSeq(
    'passphrase',
    'passphrase' / SshBytes,
    Terminated
)

# define and parse the size field separately in order to have a way to know how
# many bytes to expect to read on the socket.
SshAgentRequestHeader = Int32ub
SshAgentRequest = FocusedSeq(
    'request',
    'request' / Prefixed(Int32ub, Struct(
        'code' / PyEnum(Byte, SshAgentRequestCode),
        'message' / Switch(
            this.code, {
                SshAgentRequestCode.REQUEST_IDENTITIES: Terminated,
                SshAgentRequestCode.ADD_IDENTITY: SshAddIdentity,
                SshAgentRequestCode.REMOVE_IDENTITY: SshRemoveIdentity,
                SshAgentRequestCode.REMOVE_ALL_IDENTITIES: Terminated,
                SshAgentRequestCode.SIGN_REQUEST: SshAgentSignRequest,
                SshAgentRequestCode.LOCK: SshAgentLock,
                SshAgentRequestCode.UNLOCK: SshAgentUnlock,
            },
            default=GreedyBytes
        ),
        Terminated
    )),
    Terminated
)

SshAgentResponseHeader = Int32ub
SshAgentResponse = FocusedSeq(
    'response',
    'response' / Prefixed(Int32ub, Struct(
        'code' / PyEnum(Byte, SshAgentResponseCode),
        'message' / Switch(
            this.code, {
                SshAgentResponseCode.FAILURE: Terminated,
                SshAgentResponseCode.SUCCESS: Terminated,
                SshAgentResponseCode.IDENTITIES_ANSWER: SshAgentIdentities,
                SshAgentResponseCode.SIGN_RESPONSE: SshSignature,
                SshAgentResponseCode.EXTENSION_FAILURE: Terminated,
            },
            default=GreedyBytes
        ),
        Terminated
    )),
    Terminated
)
